Std/Tactic/TryThis.lean

```diff
@@ -1,42 +1,126 @@
+# "Try this" support
 
-Lean 4 does not yet support code actions
-to present tactic suggestions in the editor.
-This file contains a preliminary API
-that tactics can call to show suggestions.
+This implements a mechanism for tactics to print a message saying `Try this: <suggestion>`,
+where `<suggestion>` is a link to a replacement tactic. Users can either click on the link
+in the suggestion (provided by a widget), or use a code action which applies the suggestion.
 -/
 namespace Std.Tactic.TryThis
 
-open Lean Elab Elab.Tactic PrettyPrinter Meta
+open Lean Elab Elab.Tactic PrettyPrinter Meta Server Lsp RequestM
+
+/-- An info-tree data node corresponding to an application of the "Try this" command. -/
+structure TryThisInfo where
+  /-- The suggested replacement for this syntax, usually the rendering of another tactic syntax. -/
+  suggestion : String
+  /-- This is the span to replace with `suggestion`. If not supplied it will default to
+  the span of the syntax on which this info node is placed. -/
+  span? : Option (String.Pos × String.Pos) := none
+  deriving TypeName
+
+/--
+This is a code action provider that looks for `TryThisInfo` nodes and supplies a code action to
+apply the replacement.
+-/
+@[codeActionProvider] def tryThisProvider : CodeActionProvider := fun params snap => do
+  let doc ← readDoc
+  let startPos := doc.meta.text.lspPosToUtf8Pos params.range.start
+  let endPos := doc.meta.text.lspPosToUtf8Pos params.range.end
+  pure <| snap.infoTree.foldInfo (init := #[]) fun _ctx info result => Id.run do
+    let .ofCustomInfo info := info | result
+    let some tti := info.value.get? TryThisInfo | result
+    let some (head, tail) := (tti.span? <|> return (← info.stx.getPos?, ← info.stx.getTailPos?))
+      | result
+    unless head ≤ endPos && startPos ≤ tail do return result
+    result.push {
+      eager.title := "Apply 'Try this'"
+      eager.kind? := "refactor"
+      eager.edit? := WorkspaceEdit.ofTextEdit params.textDocument.uri {
+        range := doc.meta.text.utf8PosToLspRange head tail
+        newText := tti.suggestion
+      }
+    }
+
+/--
+This is a widget which is placed by `TryThis.addSuggestion`; it says `Try this: <replacement>`
+where `<replacement>` is a link which will perform the replacement.
+-/
+@[widget] def tryThisWidget : Widget.UserWidgetDefinition where
+  name := "Tactic replacement"
+  javascript := "
+import * as React from 'react';
+import { EditorContext } from '@leanprover/infoview';
+const e = React.createElement;
+export default function(props) {
+  const editorConnection = React.useContext(EditorContext)
+  function onClick() {
+    editorConnection.api.applyEdit({
+      changes: { [props.pos.uri]: [{ range: props.range, newText: props.suggestion }] }
+    })
+  }
+  return e('div', {className: 'ml1'}, e('pre', {className: 'font-code pre-wrap'}, [
+    'Try this: ',
+    e('a', {onClick, className: 'link pointer dim', title: 'Apply suggestion'}, props.suggestion),
```

Comment:
Could you add `style: { cursor: 'pointer' }` to the `onClick` structure, so that the mouse becomes a pointer on hover? (suggested by Adam Topaz in the Zulip thread)